src/crud.rs

```diff
@@ -51,152 +50,220 @@ impl EditGuard for NameGuard {
         }
     }
+    impl Events for Self {
+        type Data = Option<Entry>;
+    }
 }
 
 impl_scope! {
     #[impl_default]
-    #[widget {
-        layout = row! [
-            Button::label_msg("Create", Control::Create).map_any(),
-            self.update,
-            self.delete,
-        ];
-    }]
+    #[widget]
+    #[layout(row! [
+        Button::label_msg("Create", Control::Create).map_any(),
+        self.update,
+        self.delete,
+    ])]
     struct Controls {
         core: widget_core!(),
         #[widget(&())] update: Button<AccessLabel> = Button::label_msg("Update", Control::Update),
         #[widget(&())] delete: Button<AccessLabel> = Button::label_msg("Delete", Control::Delete),
     }
     impl Events for Self {
-        type Data = bool;
+        type Data = Option<Entry>;
 
-        fn update(&mut self, cx: &mut ConfigCx, any_selected: &bool) {
+        fn update(&mut self, cx: &mut ConfigCx, selected: &Self::Data) {
             if self.update.id_ref().is_valid() {
-                let disable = !any_selected;
+                let disable = selected.is_none();
                 cx.set_disabled(self.update.id(), disable);
                 cx.set_disabled(self.delete.id(), disable);
             }
         }
     }
 }
 
+struct EntriesClerk {
+    // Note: deleted entries are replaced with None instead of being removed.
+    // This is an easy way of ensuring that Key-Entry mappings do not change.
+    entries: Vec<Option<Entry>>,
+    filtered_entries: Vec<usize>,
+}
+
+impl DataClerk<usize> for EntriesClerk {
+    type Data = ContainsCaseInsensitive;
+    type Key = usize;
+    type Item = Entry;
+    type Token = usize;
+
+    fn update(&mut self, _: &mut ConfigCx, _: Id, filter: &Self::Data) -> DataChanges {
+        // TODO(opt) determine when updates are a no-op and return DataChanges::None
+
+        self.filtered_entries = self
+            .entries
+            .iter()
+            .enumerate()
+            .filter(|(_, opt)| {
+                opt.as_ref()
+                    .map(|entry| filter.matches(entry))
+                    .unwrap_or(false)
+            })
+            .map(|(i, _)| i)
+            .collect();
+
+        DataChanges::Any
+    }
+
+    fn len(&self, _: &Self::Data, _: usize) -> DataLen<usize> {
+        DataLen::Known(self.filtered_entries.len())
+    }
+
+    fn update_token(
+        &self,
+        _: &Self::Data,
+        index: usize,
+        _: bool,
+        token: &mut Option<usize>,
+    ) -> TokenChanges {
+        let key = self.filtered_entries.get(index).cloned();
+        if *token == key {
+            TokenChanges::None
+        } else {
+            *token = key;
+            TokenChanges::Any
+        }
+    }
+
+    fn item(&self, _: &Self::Data, key: &usize) -> &Entry {
+        self.entries
+            .get(*key)
+            .map(|inner| inner.as_ref())
+            .flatten()
+            .unwrap()
+    }
+}
```

Comment:
This is inelegant: `DataClerk` is a tedious low-level interface. The higher-level `DataGenerator` trait doesn't seem appropriate considering we can reference entries. Not very important.

We also need to use `.unwrap()` in `fn item` since the result is no longer an `Option`. Missing items here *are* a bug, but should such a bug force a panic in user-space?